python/ray/_private/async_compat.py

```diff
@@ -28,8 +28,30 @@ def try_install_uvloop():
 
 
 def is_async_func(func):
-    """Return True if the function is an async or async generator method."""
-    return inspect.iscoroutinefunction(func) or inspect.isasyncgenfunction(func)
+    """
+    Returns True if the passed function is a coroutine or an async generator.
+
+    note: we prefer this version as Python Coroutines != Cython Coroutines
+    and Cython Coroutines are not detected properly by the inspect module
+
+    this is the optimized version from the links below, using the flag as defined by
+    the inspect module
+    https://github.com/cython/cython/blob/master/tests/run/test_coroutines_pep492.pyx#L889
+    https://github.com/hugapi/hug/blob/develop/hug/introspect.py#L33
+
+    The issues above outline the PR where this was added to ensure async flag support
+    on Cython
+    https://github.com/cython/cython/pull/4902
+    """
+    is_async_cython = func.__code__.co_flags & 0x80 or getattr(
```

Comment:
I'm trying to understand the issue, so you want to define a Cython async actor?

Do you have a test code that's broken without this PR?

Reply:
Thanks for working with me on this! I indeed would love to define an Cython Async Actor and let it be detected correctly through Ray. Whenever trying this I get an error now (ref Issue)

Would you be able to provide me a pointer (first time contributing) on how I could make a test that cythonizes the required Actor implementation?

Reply:
Do you have a script showing what you are doing now?

Reply:
Currently the following is executed (where we cythonize our project first and then run the test suite):
1. Create an Actor Wrapper (for easiness sake - we can create Actors with `a = Actor(DemoClass)` so we can execute methods like `a.hello()`
2. If the Actor implementation has an Async method, it gets cythonized as well and aliased to `_cython_3_0_2.coroutine'` and an error is being thrown each time

Reply:
This is the test code I use where if the DemoClasses are cythonized it breaks whereas I expect Ray to then execute the async methods and await them

```python
import pickle
import pytest
import ray
import asyncio
# import cython

from composabl_ray.actor.actor import Actor

DEMO_CONFIG = {"config": "test"}

class DemoComplexObject:
    def __init__(self, data):
        self.data = data

class DemoNested:
    def __init__(self, config):
        self.config = config

    def nested_method_sync(self, complex_obj):
        return DemoComplexObject(complex_obj.data + " processed")

    async def nested_method_async(self, complex_obj):
        return DemoComplexObject(complex_obj.data + " processed")


class Demo:
    def __init__(self, config):
        self.config = config

        # Create a nested actor
        self.nested_actor = Actor(DemoNested, {"is_nested": True})

    def sync_void(self):
        print("Hello, world!")

    def sync_return(self):
        return self.config

    def sync_throw(self):
        raise ValueError("This is a test error")

    async def async_void(self):
        print("Hello, world!")

    async def async_return(self):
        return self.config

    async def async_throw(self):
        raise ValueError("This is a test error")

    def nested_outer_method_sync(self, complex_obj):
        return self.nested_actor.nested_method_sync(complex_obj)

    async def nested_outer_method_async(self, complex_obj):
        return await self.nested_actor.nested_method_async(complex_obj)

# def test_is_cython():
#     """
#     It should ensure 100% that this is being ran as a test suite in Cython as Ray is very flaky with actors that use Cython
#     """
#     assert cython.compiled == True

@pytest.mark.asyncio
async def test_actor_with_demo():
    ray.init(ignore_reinit_error=True)

    # Create an instance of the Actor class with the Demo class
    demo_actor = Actor(Demo, DEMO_CONFIG)

    # We should be able to call a sync method that prints
    demo_actor.sync_void()

    # We should be able to call an async method that prints
    await demo_actor.async_void()

    # We should be able to call a sync method that returns
    result = demo_actor.sync_return()
    assert result == DEMO_CONFIG

    # We should be able to call an async method that returns
    result = await demo_actor.async_return()
    assert result == DEMO_CONFIG

    # We should be able to raise an error in a sync method
    with pytest.raises(ValueError) as excinfo:
        demo_actor.sync_throw()

    assert "This is a test error" in str(excinfo.value)

    # We should be able to raise an error in an async method
    with pytest.raises(ValueError) as excinfo:
        await demo_actor.async_throw()

    assert "This is a test error" in str(excinfo.value)

    # We should be able to call a sync method on a nested actor
    result = demo_actor.nested_outer_method_sync(DemoComplexObject("test"))
    assert result.data == "test processed"

    # We should be able to call an async method on a nested actor
    result = await demo_actor.nested_outer_method_async(DemoComplexObject("test"))
    assert result.data == "test processed"

    # We should be able to pickle the actors
    actor_serialized = pickle.dumps(demo_actor)
    actor_deserialized = pickle.loads(actor_serialized)
    actor_deserialized_res = actor_deserialized.nested_outer_method_sync(DemoComplexObject("test_after_pickling"))
    assert actor_deserialized_res.data == "test_after_pickling processed"

    # Terminate the actor
    demo_actor.terminate()

    ray.shutdown()

if __name__ == "__main__":
    asyncio.run(test_actor_with_demo())
```

Reply:
As extra, I tried the following so far:

* `print(inspect.isawaitable(attr))`  # returns False
* `print(inspect.iscoroutine(attr))`  # returns False
* `print(inspect.iscoroutinefunction(attr)) ` # returns False
* `print(asyncio.iscoroutine(attr))`  # returns False
* `print(attr.__code__.co_flags & 0x0080 or getattr(attr, "_is_coroutine", False))` # returns True

Reply:
Do you have a minimal reproduce script? The above code is not runnable: `Actor` definition is missing and how do you cythonize you project.

Reply:
I will work on creating a specific project for it that I can share and share it asap. 

For cythonizing, we use a custom setup.py that compiles all the files. In short it's a `cythonize -i <file>` that we can then import

Reply:
@jjyao there you go: https://github.com/XavierGeerinck/ray-repro-cython-async

Reply:
<img width="1146" alt="image" src="https://github.com/ray-project/ray/assets/591282/a23d3111-b12c-4d12-85b8-e20367166839">

Reply:
any update on this? @jjyao

Reply:
cc @jjyao

Reply:
any update on this? @jjyao